File: C++/03_Greedy/큰수의법칙.py
import sys

n, m, k = map(int, sys.stdin.readline().split())
data = list(map(int, sys.stdin.readline().split()))
data.sort()
res = 0
if m < k:                   # m 이 k 보다 작은 경우 그냥 m만큼 제일 큰 수를 곱하게 종료
    res = m * data[-1]
else:
    while m > 0:           
        if m > k:               # k 가 m 보다 작은 경우 k개의 제일 큰 수와 1개의 2번째로 큰 수를 더한다
            res += k * data[-1] + data[-2]  
            m -= k + 1          # 그 이후 m 에서 k + 1개 만큼을 빼준다
        else:
            res += m * data[-1] # k보다 더 적은 숫자가 남았다면 그냥 그 수만큼 제일 큰 수르 더한다
print(res)

File: C++/03_Greedy/큰수의법칙.cpp
#include <iostream>
#include <algorithm>

using namespace std;
int main() {
    int n, m, k, tmp = 0, res = 0;
    cin >> n >> m >> k;
    int * data = new int[n];
    for(int i = 0; i < n; i ++)
    {
        cin >> tmp;
        data[i] = tmp;
    }
    sort(data, data + n);           
    if (m < k)
    {
        res = m * data[n - 1];
    }
    else
    {
        while (m > 0)
        {
            if (m  >  k)
            {
                res += k * data[n - 1] + data[n - 2];
                m -= k + 1;
            }
            else
            {
                res += m * data[n - 1];
                break;
            }
        }
    }
    cout << res;
}

//  sort(arr, arr+n);
//  sort(v.begin(), v.end());sort(v.begin(), v.end(), compare);     사용자 정의 함수 사용
//  sort(v.begin(), v.end(), greater<자료형>());               내림차순 (Descending order)
//  sort(v.begin(), v.end(), less<자료형>                      오름차순 (default = Ascending order)
